src/zkp/pimod.rs

```diff
@@ -97,29 +111,35 @@ impl Proof for PiModProof {
 
         Self::fill_transcript(transcript, context, input, &w)?;
 
-        let mut elements = vec![];
-        for _ in 0..LAMBDA {
-            let y = positive_challenge_from_transcript(transcript, &input.N)?;
-            let (a, b, x) = y_prime_combinations(&w, &y, &secret.p, &secret.q)?;
-
-            // Compute phi(N) = (p-1) * (q-1)
-            let phi_n = (&secret.p - 1) * (&secret.q - 1);
-            let exp = input.N.invert(&phi_n).ok_or_else(|| {
-                error!("Could not invert N");
-                InternalError::InternalInvariantFailed
-            })?;
-            let z = modpow(&y, &exp, &input.N);
-
-            elements.push(PiModProofElements {
-                x: x[0].clone(),
-                a,
-                b,
-                z,
-                y,
-            });
-        }
-
-        let proof = Self { w, elements };
+        let elements: Result<Vec<PiModProofElements>> = (0..LAMBDA)
+            .map(|_| {
+                positive_challenge_from_transcript(transcript, &input.N).and_then(|y| {
```

Comment:
On second look, I find this hard enough to read that I am changing this from a suggestion to a todo. Instead of
```
positive_challenge_from_transcript(...).and_then(|y| {
    // rest of the code...
}
```
I would like you to change it to
```
let y = positive_challenge_from_transcript(...)?;
// rest of the code...
```